File: leetcode/201-300/281-zigzag-iterator/bench/zigzag.py
#!/usr/bin/env python3
"""LeetCode 281 bench mirror — Python. Same cursor iterator, same skip scan."""
K = 64
ROUNDS = 2200


def drain_sink(lists):
    cursor = [0] * K
    remaining = sum(len(v) for v in lists)
    turn = tried = 0
    while tried < K and cursor[turn] >= len(lists[turn]):
        turn = (turn + 1) % K
        tried += 1
    h, pos = 0, 1
    while remaining > 0:
        t = turn
        v = lists[t][cursor[t]]
        cursor[t] += 1
        remaining -= 1
        h = (h * 31 + v * pos) % 1000000007
        pos += 1
        turn = (t + 1) % K
        scan = 0
        while scan < K and cursor[turn] >= len(lists[turn]):
            turn = (turn + 1) % K
            scan += 1
    return h


def main():
    seed = 20260819
    lists = []
    for _ in range(K):
        seed = (seed * 1103515245 + 12345) % 2147483648
        ln = 1 + (seed // 7) % 2000
        v = []
        for _ in range(ln):
            seed = (seed * 1103515245 + 12345) % 2147483648
            v.append(seed % 100003)
        lists.append(v)
    sink = 0
    for _ in range(ROUNDS):
        sink = (sink + drain_sink(lists)) % 1000000007
    print(sink)


main()

File: leetcode/201-300/281-zigzag-iterator/bench/zigzag.rs
// LeetCode 281 bench mirror — Rust. Same cursor iterator, same skip scan.
const K: usize = 64;
const ROUNDS: usize = 2200;
fn drain_sink(lists: &[Vec<i64>]) -> i64 {
    let mut cursor = [0usize; K];
    let mut remaining: i64 = lists.iter().map(|v| v.len() as i64).sum();
    let mut turn = 0usize;
    let mut tried = 0usize;
    while tried < K && cursor[turn] >= lists[turn].len() { turn = (turn + 1) % K; tried += 1; }
    let (mut h, mut pos) = (0i64, 1i64);
    while remaining > 0 {
        let t = turn;
        let v = lists[t][cursor[t]];
        cursor[t] += 1; remaining -= 1;
        h = (h * 31 + v * pos) % 1000000007; pos += 1;
        turn = (t + 1) % K;
        let mut scan = 0usize;
        while scan < K && cursor[turn] >= lists[turn].len() { turn = (turn + 1) % K; scan += 1; }
    }
    h
}
fn main() {
    let mut seed: i64 = 20260819;
    let mut lists: Vec<Vec<i64>> = Vec::with_capacity(K);
    for _ in 0..K {
        seed = (seed*1103515245+12345)%2147483648;
        let len = 1 + (seed/7) % 2000;
        let mut v = Vec::with_capacity(len as usize);
        for _ in 0..len { seed = (seed*1103515245+12345)%2147483648; v.push(seed % 100003); }
        lists.push(v);
    }
    let mut sink = 0i64;
    for _ in 0..ROUNDS { sink = (sink + drain_sink(&lists)) % 1000000007; }
    println!("{}", sink);
}
